src/algorithms/calorimetry/CalorimeterClusterRecoCoG.cc

```diff
@@ -338,4 +295,169 @@ std::optional<edm4eic::MutableCluster> CalorimeterClusterRecoCoG::reconstruct(co
   return std::move(cl);
 }
 
+//------------------------------------------------------------------------
+void CalorimeterClusterRecoCoG::associate(
+  const edm4eic::Cluster& cl,
+#if EDM4EIC_VERSION_MAJOR >= 7
+  const edm4eic::MCRecoCalorimeterHitAssociationCollection* mchitassociations,
+#else
+  const edm4hep::SimCalorimeterHitCollection* mchits,
+#endif
+  edm4eic::MCRecoClusterParticleAssociationCollection* assocs
+) const {
+
+  // --------------------------------------------------------------------------
+  // Association Logic
+  // --------------------------------------------------------------------------
+  /*  1. identify all sim hits associated with a given protocluster, and sum
+   *     the energy of the sim hits.
+   *  2. for each sim hit, identify all contributing primary particles, and sum
+   *     their contributed energy.
+   *  3. create an association for each contributiong primary with a weight
+   *     of contributed energy over total energy.
+   */
+
+  // bookkeeping maps for associated primaries
+  std::map<std::size_t, std::pair<std::size_t, std::size_t>> mapMCParToSimIndices;
+  std::map<std::size_t, double> mapMCParToContrib;
+
+  // --------------------------------------------------------------------------
+  // 1. get associated sim hits and sum energy
+  // --------------------------------------------------------------------------
+  double eSimHitSum = 0.;
+  for (std::size_t iHit = 0; auto clhit : cl.getHits()) {
+
+    // vector to hold associated sim hits
+    std::vector<std::size_t> vecAssocSimHits;
```

Comment:
Code with indices is incomprehensible. Just make a vector of those proxy objects
```suggestion
    std::vector<edm4eic::MCRecoCalorimeterHitAssociation> vecAssocSimHits;
```

Reply:
I'm not opposed to going about it this way (it would definitely simplify the bookkeeping), but I think this actually touches on the discussion in #1600 a little:
https://github.com/eic/EICrecon/pull/1600#discussion_r1733984818

Would there be an issue with working with the user-layer objects in this case?

Reply:
Yes, this is the intended way.